hw/drivers/lp5523/src/lp5523_shell.c

```diff
@@ -0,0 +1,1017 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <string.h>
+#include <errno.h>
+#include "sysinit/sysinit.h"
+#include "console/console.h"
+#include "shell/shell.h"
+#include "hal/hal_gpio.h"
+#include "lp5523/lp5523.h"
+#include "parse/parse.h"
+
+#if MYNEWT_VAL(LP5523_CLI)
+
+static int lp5523_shell_cmd(int argc, char **argv);
+
+static struct shell_cmd lp5523_shell_cmd_struct = {
+    .sc_cmd = "lp5523",
+    .sc_cmd_func = lp5523_shell_cmd
+};
+
+static struct sensor_itf *lp5523_itf;
```

Comment:
lp5523_itf should be defined here, something like below in hw/drivers/sensors/bno055/src/bno055_shell.c:

```
   45 static struct sensor_itf g_sensor_itf = {
   46     .si_type = MYNEWT_VAL(BNO055_SHELL_ITF_TYPE),
   47     .si_num = MYNEWT_VAL(BNO055_SHELL_ITF_NUM),
   48     .si_addr = MYNEWT_VAL(BNO055_SHELL_ITF_ADDR),
   49 };
```
where BNO055_SHELL_ITF_TYPE, BNO055_SHELL_ITF_NUM and BNO055_SHELL_ITF_ADDR would be syscfgs defined in the drivers syscfg.yml file.

Reply:
I agree that I should follow the convention here, but it seems odd that one should have to have 2 different structures in memory with the same constant value (one to pass into `os_dev_create` and another that is declared static in the `*_shell.c` file). Is it the intention to include the `*_shell_init` functions in sysinit at some point in the future?

I would love to see values pulled from the config similar to what you posted above in `sensor_creator.c`. It would them be simple to unify the 2 `sensor_itf` structs and bring the shell initialisation into sysinit while allowing the user to pick their own interface without changing any of the code in the repo, something like:

```
MYNEWT_VAL(BNO055_OFB)
static struct sensor_itf creator_itf_bno = {
    .si_type = MYNEWT_VAL(BNO055_CREATOR_ITF_TYPE),
    .si_num  = MYNEWT_VAL(BNO055_CREATOR_ITF_NUM),
    /* HW I2C address for the BNO055 */
    .si_addr = MYNEWT_VAL(BNO055_CREATOR_ITF_ADDR),
};
#endif

#if MYNEWT_VAL(BNO055_OFB)
    rc = os_dev_create((struct os_dev *) &bno055, "bno055_0",
      OS_DEV_INIT_PRIMARY, 0, bno055_init, (void *)&creator_itf_bno);
    assert(rc == 0);

    rc = config_bno055_sensor();
    assert(rc == 0);

#if MYNEWT_VAL(BNO055_CLI)
    rc = bno055_shell_init(&creator_itf_bno);
    assert(rc == 0);
#endif
#endif
```

Reply:
The suggestion is great and probably we want to able to do that at some point but sensor device shell can't really be initialized using sysinit. Reason is we we cannot specify arguments in sysinit and that extent of code generation is not our intent. Also, we want the app to initialize the shell only for one instance of the sensor and it's only for experimentation purposes. It is not going to be present in most products.
so, the expected flow of development would be:
- Bring up one sensor using the driver
- Use the shell to talk to it and experiment
- Test all sensor features
- Once tested, plug it into the sensor framework
- Test the sensor using the sensor shell. 
- If you see the data coming out as expected, the driver is good.
The shell is optional and it is not needed most of the times.
Hope I answered your question.